### PR TITLE
Added check of max size limit for proofs & public inputs params in VerifyProof

### PR DESCRIPTION
**Pull request by Kushal7788**
- Add zk module params for max byte sizes: Groth16 proof / public inputs (UTF-8 string length) and UltraHonk proof / public inputs (raw bytes).
- Enforce those limits in VerifyProof so oversized payloads are rejected before verification.
- Add v3 zk migration to set defaults for the new fields.
- Extend keeper/query tests and adjust related types, errors, and genesis handling.

### CODE REVIEW

File: x/zk/keeper/query_server.go [hunk 1]
```diff
@@ -31,11 +31,39 @@ func (q Querier) ProofVerify(c context.Context, req *types.QueryVerifyRequest) (
...
 	}
 
+	params, err := q.GetParams(c)
+	if err != nil {
+		return nil, err
+	}
+
 	// Validate proof bytes
 	if len(req.Proof) == 0 {
 		return nil, errors.Wrap(types.ErrInvalidRequest, "proof cannot be empty")
```
[the remark this answers is not in the document]
Comment (2xburnt) [reply x/zk/keeper/query_server.go]: this seems reasonable

**Comment by crucible-burnt** [comment]
## Security Review

### 🔴 Medium — Fix Before Merge

**Integer overflow in Groth16 public-input size accumulation** (`keeper/query_server.go:54–57`)

```go
var publicInputsSize uint64
for _, in := range req.PublicInputs {
    publicInputsSize += uint64(len(in))
}
```

`req.PublicInputs` is `[]string` with no element count limit. A large number of short strings will wrap `publicInputsSize` back below `MaxGroth16PublicInputSizeBytes`, bypassing the cap entirely. Fix:

```go
if publicInputsSize+uint64(len(in)) < publicInputsSize {
    return nil, errors.Wrap(types.ErrPublicInputsTooLarge, "overflow")
}
publicInputsSize += uint64(len(in))
```

---

### 🟡 Low

**No gas metering** — size limits reduce the surface but BN254 pairing still runs for free on max-size inputs. Add a TODO comment at minimum linking to the gas metering work.

**Silent mutation in `ValidateBasic`** — `types/msgs.go:97` calls `m.Params.WithMaxLimitDefaults().Validate()` but `SetParams` also calls `WithMaxLimitDefaults()` internally. Defaulting happens in two places — worth consolidating or adding a comment explaining the intent.

**Stale scaffolding comment** — `keeper/keeper.go` contains `// Add these getter methods to your keeper if they don't exist` — should be removed.

---

### 🔵 Test Coverage Gaps

- `types/params_test.go` — no tests for the 4 new `Validate()` error cases or `WithMaxLimitDefaults()` behavior
- `migrations/v3/migrate_test.go` — no tests for the `ErrNotFound` path, idempotency when params already non-zero, or partial-fill case

---

### ✅ Looks Good

- Migration wiring is correct — `ConsensusVersion = 3` + `RegisterMigration(2→3)` auto-runs via `RunMigrations()` in the v29 upgrade handler; no `upgrades.go` changes needed
- UltraHonk uses `[]byte` so no overflow issue on that path
- Default values (Groth16: 4KB proof / 30KB inputs; UltraHonk: 20KB proof / 10KB inputs) are exercised against real fixtures in `query_server_test.go`

**Verdict:** The overflow issue is the blocker — easy one-liner fix. Test gaps worth addressing but not blocking under time pressure. Rest is clean.

**Review by Peartes — APPROVED** [review]

**Comment by crucible-burnt** [comment]
## 🔍 Crucible Security Review

### Summary
Solid hardening PR that adds pre-verification size gates to both `ProofVerify` (Groth16) and `ProofVerifyUltraHonk` endpoints, directly closing an unbounded-input attack surface on the gRPC query path. Implementation is clean and well-tested.

### Security Assessment
- **Risk Level**: Low (minor notes below, no blockers)

**Positive findings:**
- Size checks fire **before** expensive crypto operations — correct placement; prevents resource exhaustion
- Overflow guard on Groth16 public inputs sum (`publicInputsSize+inputSize < publicInputsSize`) is correct and necessary since inputs is `[]string`
- UltraHonk public inputs are raw `[]byte`, so a single `len()` check is appropriate — no overflow risk
- New error codes 1110/1111 have no collision with existing zk error registry
- `WithMaxLimitDefaults()` applied at write time (SetParams), read time (GetParams), genesis validation, and ValidateBasic — multi-layer defense is correct given the upgrade compat requirement

**Notes:**
- `x/zk/keeper/query_server.go:306-308` — New early-return check `if req.VkeyName == "" && req.VkeyId == 0` is a behavioral change for existing callers; any client sending empty vkey fields previously got a later error, now gets `ErrInvalidRequest` before size checks. Not a security issue, but worth flagging in release notes.
- `x/zk/types/params.go:Validate()` only rejects `== 0` for the new fields. There is no lower-bound sanity check, so governance could technically set `max_groth16_proof_size_bytes = 1`, which would break all ZK verification for legitimate users. Consider documenting minimum safe values or adding a floor (e.g., ≥ 1 KiB). This is a governance hygiene concern, not an exploitable vulnerability.
- `x/zk/types/msgs.go:921` — `WithMaxLimitDefaults()` is applied inside `ValidateBasic()` before `Validate()`, which means a `MsgUpdateParams` with `MaxGroth16ProofSizeBytes = 0` silently gets backfilled to default rather than rejected. Intentional for upgrade compat, but this is a subtle behavior — governance tooling that submits params with explicit zeros may not notice they're being overridden. Consider at minimum a log line or consider rejecting explicit zeros post-migration.

### Immunefi Pattern Check
- Directly addresses the unbounded memory allocation / gRPC query DoS class. The pre-verification size gates are precisely the mitigation for that class of report. No new patterns introduced.
- The `jwk/query.pb.go` doc changes (adding `compact serialization format` comment to `SigBytes`, deprecating `ValidateJWT`) are consistent with hardening work on the JWT/JWS validation path. The compact serialization note is accurate and reduces false-report risk on that endpoint.

### False Report Risk
- **Low.** The `WithMaxLimitDefaults()` on multiple layers might look like a "bypass" to an auditor (e.g., "you can store zero and the check is skipped"), but the read-time backfill in `GetParams()` closes that. The enforcement path is airtight.
- The comment at `query_server.go:324` ("Approximate public-input payload size as total UTF-8 byte length") could invite a report claiming UTF-8 multi-byte characters are under-counted. This is accurate — `len(string)` in Go returns byte length, not rune count — but the comment wording "UTF-8 byte length" is correct and should deter this.

### Code Quality Notes
- Test coverage is thorough: param-enforced limits, boundary conditions (exact max passes, max+1 fails), migration idempotency, and the `fills only missing limit fields` case
- `TestQueryProofVerify` error-matching by `tc.name` string switch (query_server_test.go:426-431) is fragile — test names are the key, so a rename would silently break the assertion. Low-severity but worth a follow-up refactor
- Default limit values (Groth16 proof: 4 KiB, public inputs: 30 KiB, UltraHonk proof: 20 KiB, public inputs: 10 KiB) — are these sized against real circuit outputs? Would be good to have a comment referencing the expected production proof sizes to justify these values, otherwise an auditor will ask

### Status
No blockers. The two behavioral notes (vkey empty-field check and `WithMaxLimitDefaults` in ValidateBasic) are worth documenting. Recommend confirming the default limit values against real circuit benchmarks before merge. Governance floor consideration is optional but worth a follow-up issue.
